### ISSUE TITLE
Helm collector times out with lots of charts 

### ISSUE THREAD

**Issue by huang-jy** [doitintl/kube-no-trouble#358]
When running kubent on an installation which has a lot of helm2 charts released (we have 153 on one of our clusters) , it errors:

```
11:45AM ERR Failed to retrieve data from collector error="stream error when reading response body, may be caused by closed connection. Please retry. Original error: stream error: stream ID 67; INTERNAL_ERROR" name="Helm v2"
```

Disabling the collector with `--helm2=false` will remove the error, but you lose all the benefit of the tool as it no longer checks the charts.

You _can_ work around it by extracting the helm list then iterating it through kubent:

```shell
for chart in $(helm2 ls -aq )
do
  helm2 get manifest $chart | kubent -c=false --helm2=false --helm3=false -t xx.yy -f -
done
```

**Comment by nicholas-marchini** [comment]
We also have this issue

**Comment by mikebell90** [comment]
Also true of helm 3. I think it's basically a large set of secrets tends to make things bork. To be fair, a similar tool, pluto has the same issue, so guessing it's in a shared helm api library

**Comment by huang-jy** [comment]
Edited title since it seems to affect both helm2 and helm3 collectors

**Comment by mikebell90** [comment]
Yeah here's my semi working workaorund

`#!/bin/sh -e
echo "Note: Run until it's clean. Will exit automatically everytime a bad api is found."
kubent -e --helm3=false --helm2=false
for ns in $(kubectl get ns -o custom-columns=":metadata.name")
do
	echo namespace $ns
	for chart in $(helm -n $ns list -a -q)
	do
  		echo Chart $chart
  		helm -n $ns get manifest $chart | kubent -e -c=false --helm2=false --helm3=false -f -
	done

done


`

**Comment by matandomuertos** [comment]
> Yeah here's my semi working workaorund
> 
> ```
> #!/bin/sh -e
> echo "Note: Run until it's clean. Will exit automatically everytime a bad api is found."
> kubent -e --helm3=false --helm2=false
> for ns in $(kubectl get ns -o custom-columns=":metadata.name")
> do
> 	echo namespace $ns
> 	for chart in $(helm -n $ns list -a -q)
> 	do
>   		echo Chart $chart
>   		helm -n $ns get manifest $chart | kubent -e -c=false --helm2=false --helm3=false -f -
> 	done
> 
> done
> ```

Thanks, that fixed my issue, I did a small change to be able to send additional flags to kubent, nothing fancy but I needed `-t x.yy`: https://gist.github.com/matandomuertos/0c49799d18b37a8a5bc8b2939bc842c4

Thanks!

**Comment by mikebell90** [comment]
No problem. Of course, the hack of going namespace by namespace is to reduce the overally load of the helm library.

TBH Helm has a very very bad and unscalable design with its default secret store. There's no index IIRC, it's pretty much just iterating all the secrets of which there is one per revision per release. Arguably if they required US ASCII (I know old school) and stored 36 secrets (A-Z0-9, case insenstive), they might fit an index, but they probably would have concurrency issue.

**Comment by mikebell90** [comment]
Again I want to make the point, this is almost certainly not an issue for this project. It's the helm implementation.

**Comment by huang-jy** [comment]
Could we add a `--timeout` switch to allow us to wait longer than normal? Sure the helm implementation isn't perfect, but we can either work around it by looping through the helm charts or allow kubent to do that but allow it to wait longer?

**Comment by mikebell90** [comment]
interesting question, but I don't know if the helm apis let you do that.

**Comment by huang-jy** [comment]
Well, `helm ls` doesn't have a timeout switch so I don't think it has a timeout, but since `helm ls` doesn't timeout even with lots of charts in the cluster (it can take up to 2 minutes to return on our cluster with helm2), I can only assume the timeout is happening within kubent or the library/libraries it uses.

**Comment by mikebell90** [comment]
try it while scanning ALL namespaces. which is what kubent is doing ;)

**Comment by mikebell90** [comment]
bell ~/git/puppet-modules [sigh.exerciser]$ $ helm list -A
pError: list: failed to list: stream error when reading response body, may be caused by closed connection. Please retry. Original error: stream error: stream ID 3; INTERNAL_ERROR

**Comment by huang-jy** [comment]
I've never had that output (but then again I don't have that many releases under helm3.) 

On helm2 (which doesn't namespace its releases in the listing, and therefore `-A` isn't applicable), I've had no timeouts, even with well over 150 releases. So if it doesn't timeout with a plain `ls`, `kubent` shouldn't timeout getting the same information.

Sidenote: I have noticed that under both helm2 and helm3, there's `--max-int` and `--offset` which seems to allow paginated listing. Would that be of use?

Reference:
* https://v2.helm.sh/docs/helm/#helm-list
* https://helm.sh/docs/helm/helm_list/

**Comment by stepanstipl** [comment]
Thanks everyone for helping with this! 👍 👏

As to Helm v2 - we just dropped its support (#401) (sorry to anyone using it, but it's been deprecated and unmaintained for a long while and blocking us from fixing several auth issues, which became more pressing with the new auth API).

Helm v3 - I would like to fix this if possible. We pass the K8S client to the Helm driver, so while I'll need to look into this further, I think it could be possible to increase the timeout (or even make it configurable). Having said that, it might be out of our control in some situations (some proxy/LB can be timing out depending on how K8S API is deployed). We can consider pagination if that would help.

In any case - what would be helpful at this stage - do we have a reliable way of reproducing this? (simple shell for loop deploying many helm releases, perhaps?)

[^1]: https://github.com/doitintl/kube-no-trouble/blob/d1bb4e5fd6550b533b2013671aa8419d923ee042/pkg/collector/helm3.go#L45

**Comment by huang-jy** [comment]
The best way to reproduce it I presume would be to loop a chart install so it installs as a different name and different namespace. Something as simple as 

```
for a in {1..255}
do
  helm install -n nginx-$a nginx-$a nginx/nginx
done
```

Then when its done (you'll have 255 installs), you should be able to see if that will time out the helm list

**Comment by moredatapls** [comment]
Not sure why this has been closed, but I am on 0.7.3, and I am facing the same issues with helm3. Disabling the helm3 collector works, though.

With helm3:

```bash
$> kubent -t 1.35              
12:04PM INF >>> Kube No Trouble `kubent` <<<
12:04PM INF version 0.7.3 (git sha 57480c07b3f91238f12a35d0ec88d9368aae99aa)
12:04PM INF Initializing collectors and retrieving data
12:04PM INF Target K8s version is 1.35.0
12:04PM INF Retrieved 157 resources from collector name=Cluster
12:05PM ERR Failed to retrieve data from collector error="list: failed to list: stream error when reading response body, may be caused by closed connection. Please retry. Original error: stream error: stream ID 83; INTERNAL_ERROR; received from peer" name="Helm v3"
```

Without helm3:

```bash
$> kubent -t 1.35 --helm3=false
12:06PM INF >>> Kube No Trouble `kubent` <<<
12:06PM INF version 0.7.3 (git sha 57480c07b3f91238f12a35d0ec88d9368aae99aa)
12:06PM INF Initializing collectors and retrieving data
12:06PM INF Target K8s version is 1.35.0
12:06PM INF Retrieved 157 resources from collector name=Cluster
12:06PM INF Loaded ruleset name=custom.rego.tmpl
12:06PM INF Loaded ruleset name=deprecated-1-16.rego
12:06PM INF Loaded ruleset name=deprecated-1-22.rego
12:06PM INF Loaded ruleset name=deprecated-1-25.rego
12:06PM INF Loaded ruleset name=deprecated-1-26.rego
12:06PM INF Loaded ruleset name=deprecated-1-27.rego
12:06PM INF Loaded ruleset name=deprecated-1-29.rego
12:06PM INF Loaded ruleset name=deprecated-1-32.rego
12:06PM INF Loaded ruleset name=deprecated-future.rego
```

@stepanstipl could you maybe reopen the issue? It might have been closed prematurely/by accident.
